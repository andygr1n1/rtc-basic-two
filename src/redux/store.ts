import { todoReducer } from '@/components/todo.slice'
import { configureStore } from '@reduxjs/toolkit'

export const rootStore = configureStore({
    reducer: {
        todo$: todoReducer,
    },
})

// Infer the `RootState` and `AppDispatch` types from the store itself
export type IRootStore = ReturnType<typeof rootStore.getState>

// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type IAppDispatch = typeof rootStore.dispatch
