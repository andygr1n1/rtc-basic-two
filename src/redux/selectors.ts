// reselect -> memoization
import { createSelector } from '@reduxjs/toolkit'
import { IRootStore } from './store'
import { ITodoSlice } from '@/components/todo.slice'
import { ITodo } from '@/interfaces/ITodo.interface'

export const getTodo$: (store: IRootStore) => ITodoSlice = (store) => store.todo$
export const getAllTodos: (store: IRootStore) => ITodo[] = (store) => store.todo$.todos
export const getActiveTodosFilter: (store: IRootStore) => string = (store) => store.todo$.filter

export const getTodosByFilter = createSelector(
    [getAllTodos, getActiveTodosFilter],
    (allTodos, activeFilter): ITodo[] | undefined => {
        if (activeFilter === 'all') return allTodos
        if (activeFilter === 'completed') return allTodos.filter((todo) => todo.completed)
        if (activeFilter === 'active') return allTodos.filter((todo) => !todo.completed)
    },
)

export const TodoListSelector = createSelector([getTodosByFilter, getTodo$], (getTodosByFilter, getTodo$) => {
    return { todos: getTodosByFilter, error: getTodo$.error, loading: getTodo$.loading }
})
