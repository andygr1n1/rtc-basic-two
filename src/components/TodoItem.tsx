import { ITodo } from '@/interfaces/ITodo.interface'
import { useAppDispatch } from '@/redux'
import { deleteTodoAsync, toggleTodoAsync } from './todo.slice'

export const TodoItem: React.FC<{ todo: ITodo }> = ({ todo }) => {
    const dispatch = useAppDispatch()

    const { id } = todo

    const onChange = () => {
        dispatch(toggleTodoAsync(id))
    }

    const onDelete = () => {
        dispatch(deleteTodoAsync(id))
    }

    return (
        <li className='flex gap-5'>
            <input type='checkbox' checked={todo.completed} onChange={onChange} className='peer' />
            <div className='peer-checked:line-through'>{todo.title}</div>

            <span onClick={onDelete} className='material-icons-round cursor-pointer hover:!text-red-500'>
                delete
            </span>
        </li>
    )
}
