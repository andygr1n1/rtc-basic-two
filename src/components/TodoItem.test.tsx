import { TodoItem } from './TodoItem'
import { beforeEach, describe, expect, it, afterEach, vitest } from 'vitest'
import { act, screen, fireEvent } from '@testing-library/react'
import ReactDOM from 'react-dom/client'
import * as reduxHooks from '../redux/hooks'
import * as actions from './todo.slice'

vitest.mock('react-redux')

const mockedAppDispatch = vitest.spyOn(reduxHooks, 'useAppDispatch')

let container: Element | DocumentFragment

describe('Todolist', () => {
    beforeEach(() => {
        container = document.createElement('div')
        document.body.appendChild(container)
    })
    afterEach(() => {
        container && document.body.removeChild(container)
    })

    it('should render Todo Item', () => {
        const todo = {
            id: '8',
            title: 'quo adipisci enim quam ut ab',
            completed: true,
        }

        const renderedTodo = act(() => {
            ReactDOM.createRoot(container).render(<TodoItem todo={todo} />)
        })

        const Todo = container.querySelector('li')

        expect(Todo?.tagName).toBe('LI')

        mockedAppDispatch.mockResolvedValue(vitest.fn)

        expect(renderedTodo).toMatchSnapshot()
    })

    it('should dispatch Todo actions', () => {
        const todo = {
            id: '8',
            title: 'quo adipisci enim quam ut ab',
            completed: false,
        }

        const dispatch = vitest.fn()

        mockedAppDispatch.mockReturnValue(dispatch)

        const mockedToggleComplete = vitest.spyOn(actions, 'toggleTodoAsync')

        act(() => {
            ReactDOM.createRoot(container).render(<TodoItem todo={todo} />)
        })

        fireEvent.click(screen.getByRole('checkbox'))

        expect(dispatch).toHaveBeenCalled()
        expect(dispatch).toHaveBeenCalledTimes(1)
        expect(mockedToggleComplete).toHaveBeenCalledTimes(1)
        expect(mockedToggleComplete).toBeCalledWith('8')
    })
})
