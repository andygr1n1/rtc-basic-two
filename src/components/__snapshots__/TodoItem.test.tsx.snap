// Vitest Snapshot v1

exports[`Todolist > should render Todo Item 1`] = `
{
  "then": [Function],
}
`;

exports[`Todolist > should render Todo Item 2`] = `
{
  "then": [Function],
}
`;
